### PR TITLE
[AutoML] Pull out Code Gen as separate library plus some changes in CodeGen

### PR DESCRIPTION
**Pull request by LittleLittleCloud**
We are excited to review your PR.

So we can do the best job, please check:

- [x] There's a descriptive title that will make sense to other developers some time from now. 
- [x] There's associated issues. All PR's should have issue(s) associated - unless a trivial self-evident change such as fixing a typo. You can use the format `Fixes #nnnn` in your description to cause GitHub to automatically close the issue(s) when your PR is merged.
- [x] Your change description explains what the change does, why you chose your approach, and anything else that reviewers should know.
- [x] You have included any necessary tests in the same PR.

### Related Issue
- [[Implement] Code generation improvements](https://github.com/dotnet/machinelearning-modelbuilder/issues/194)

- [CLI Hand Off](https://github.com/dotnet/machinelearning-modelbuilder/issues/128)

### CODE REVIEW

**Comment by LittleLittleCloud** [comment]
[x] pull out CodeGen
[x] use CodeGen in MLNet cli
[ ] create private feed for Model Builder on CodeGen

File: Microsoft.ML.AutoML.sln [hunk 1]
```diff
@@ -1,6 +1,6 @@
 Microsoft Visual Studio Solution File, Format Version 12.00
-# Visual Studio 15
-VisualStudioVersion = 15.0.28010.2050	
+# Visual Studio Version 16
```
Comment (JakeRadMSFT) [inline Microsoft.ML.AutoML.sln]: Undo this change. It should be able to be build on Dev 15 and Dev 16.

File: src/Microsoft.ML.CodeGen/Microsoft.ML.CodeGen.csproj [hunk 2]
```diff
@@ -0,0 +1,116 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
```
Comment (JakeRadMSFT) [inline src/Microsoft.ML.CodeGen/Microsoft.ML.CodeGen.csproj]: We need to add signing stuff in here. Let's take a look at what the other project did for signing.

File: Microsoft.ML.AutoML.sln (hunk 1 above)
Comment (LittleLittleCloud) [reply Microsoft.ML.AutoML.sln]: K

File: src/Microsoft.ML.CodeGen/CodeGenerator/CSharp/CodeGeneratorSettings.cs [hunk 3]
```diff
@@ -0,0 +1,22 @@
+﻿using Microsoft.ML.AutoML;
```
Comment (srsaggam) [inline src/Microsoft.ML.CodeGen/CodeGenerator/CSharp/CodeGeneratorSettings.cs]: nit: file headers

File: src/Microsoft.ML.CodeGen/Templates/Console/PredictProgram.cs [hunk 4]
```diff
@@ -13,7 +13,7 @@ namespace Microsoft.ML.CLI.Templates.Console
     using System.Text;
     using System.Text.RegularExpressions;
     using System.Collections.Generic;
-    using Microsoft.ML.CLI.Utilities;
+    using Microsoft.ML.CodeGen.Utilities;
```
Comment (srsaggam) [inline src/Microsoft.ML.CodeGen/Templates/Console/PredictProgram.cs]: Something seems off here. There is no corresponding change in the tt files.

Comment (srsaggam) [reply src/Microsoft.ML.CodeGen/Templates/Console/PredictProgram.cs]: General rule of thumb. Do not explicitly modify any .cs files associated with templates. They should be always auto-generated.

Comment (srsaggam) [reply src/Microsoft.ML.CodeGen/Templates/Console/PredictProgram.cs]: Or even better. have the files generated on build of the project

File: src/Microsoft.ML.CodeGen/Microsoft.ML.CodeGen.csproj (hunk 2 above)
Comment (LittleLittleCloud) [reply src/Microsoft.ML.CodeGen/Microsoft.ML.CodeGen.csproj]: Just ask @Dmitry-A , AutoML uses a different signing mechanism. So we don't have to explicitly handle that. (correct me if anything wrong)

File: src/Microsoft.ML.CodeGen/CodeGenerator/CSharp/CodeGenerator.cs [hunk 5]
```diff
@@ -10,11 +10,11 @@
 using Microsoft.CodeAnalysis;
 using Microsoft.ML.AutoML;
 using Microsoft.ML.CLI.Templates.Console;
-using Microsoft.ML.CLI.Utilities;
+using Microsoft.ML.CodeGen.Utilities;
 
-namespace Microsoft.ML.CLI.CodeGenerator.CSharp
+namespace Microsoft.ML.CodeGen.CSharp
```
[the remark this answers is not in the document]
Comment (LittleLittleCloud) [reply src/Microsoft.ML.CodeGen/CodeGenerator/CSharp/CodeGenerator.cs]: the `CodeGen` name is based on this [discussion](https://github.com/dotnet/machinelearning-modelbuilder/issues/186)

And I guess we'd better to keep namespace and dll name the same?

File: src/Microsoft.ML.CodeGen/Assembly.cs [hunk 6]
```diff
@@ -0,0 +1,10 @@
...
+
+[assembly: InternalsVisibleTo("Microsoft.ML.CodeGen.Test, PublicKey=002400000480000094000000060200000024000052534131000400000100010007d1fa57c4aed9f0a32e84aa0faefd0de9e8fd6aec8f87fb03766c834c99921eb23be79ad9d5dcc1dd9ad236132102900b723cf980957fc4e177108fc607774f29e8320e92ea05ece4e821c0a5efe8f1645c4c0c93c1ab99285d622caa652c1dfad63d745d6f2de5f17e5eaf0fc4963d261c8a12436518206dc093344d5ad293")]
+[assembly: InternalsVisibleTo("Microsoft.ML.ModelBuilder, PublicKey=002400000480000094000000060200000024000052534131000400000100010007d1fa57c4aed9f0a32e84aa0faefd0de9e8fd6aec8f87fb03766c834c99921eb23be79ad9d5dcc1dd9ad236132102900b723cf980957fc4e177108fc607774f29e8320e92ea05ece4e821c0a5efe8f1645c4c0c93c1ab99285d622caa652c1dfad63d745d6f2de5f17e5eaf0fc4963d261c8a12436518206dc093344d5ad293")]
+[assembly: InternalsVisibleTo("mlnet, PublicKey=00240000048000009400000006020000002400005253413100040000010001004b86c4cb78549b34bab61a3b1800e23bfeb5b3ec390074041536a7e3cbd97f5f04cf0f857155a8928eaa29ebfd11cfbbad3ba70efea7bda3226c6a8d370a4cd303f714486b6ebc225985a638471e6ef571cc92a4613c00b8fa65d61ccee0cbe5f36330c9a01f4183559f1bef24cc2917c6d913e3a541333a1d05d9bed22b38cb")]
```
[the remark this answers is not in the document]
Comment (LittleLittleCloud) [reply src/Microsoft.ML.CodeGen/Assembly.cs]: Yeah theoretically it shouldn't, but the `CodeGenerator` class is internal, because it requires accessibility to `pipeline` class. Probably this will no longer last after `CodeGen` relies on some intermediate output from AutoML instead of class.

File: src/Microsoft.ML.CodeGen/Microsoft.ML.CodeGen.csproj [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+    <TargetFramework>netstandard2.0</TargetFramework>
+    <DelaySign>false</DelaySign>
+    <LangVersion>7.1</LangVersion>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="Microsoft.CodeAnalysis" Version="2.10.0" />
+    <PackageReference Include="NLog" Version="4.5.11" />
+    <PackageReference Include="System.CodeDom" Version="4.5.0" />
+    <PackageReference Include="System.CommandLine.Experimental" Version="0.2.0-alpha.19174.3" />
```
[the remark this answers is not in the document]
Comment (LittleLittleCloud) [reply src/Microsoft.ML.CodeGen/Microsoft.ML.CodeGen.csproj]: Will check

File: src/Microsoft.ML.CodeGen/CodeGenerator/CSharp/CodeGenerator.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (LittleLittleCloud) [reply src/Microsoft.ML.CodeGen/CodeGenerator/CSharp/CodeGenerator.cs]: namespace and DLL for code generator will be `Microsoft.ML.CodeGenerator.`

**Comment by JakeRadMSFT** [comment]
1. Pull it out into separate library and update namespace ... make no other changes
2. Make CodeGenerator class public so that CLI doesn't need internals visible to CodeGenerator. 
3 - 5. ​Other Code Gen Improvements

File: src/Microsoft.ML.CodeGen/Assembly.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (Dmitry-A) [reply src/Microsoft.ML.CodeGen/Assembly.cs]: Submitted https://github.com/dotnet/machinelearning/issues/4129 on this.

File: src/Microsoft.ML.CodeGen/CodeGenerator/CSharp/CodeGenerator.cs [hunk 8]
```diff
@@ -282,6 +310,7 @@ private static string GeneratPredictProjectFileContent(string namespaceValue, bo
 
         private string GeneratePredictProgramCSFileContent(string namespaceValue)
         {
+            var features = _columnInferenceResult.TextLoaderOptions.Columns;
```
Comment (Dmitry-A) [inline src/Microsoft.ML.CodeGen/CodeGenerator/CSharp/CodeGenerator.cs]: This is confusing, two features variables that vary just by case of the first letter.

File: src/Microsoft.ML.CodeGen/CodeGenerator/CSharp/CodeGenerator.cs [hunk 9]
```diff
@@ -293,6 +322,8 @@ private string GeneratePredictProgramCSFileContent(string namespaceValue)
                 Separator = _columnInferenceResult.TextLoaderOptions.Separators.FirstOrDefault(),
                 AllowQuoting = _columnInferenceResult.TextLoaderOptions.AllowQuoting,
                 AllowSparse = _columnInferenceResult.TextLoaderOptions.AllowSparse,
+                Features = features.Where((str) => str.Name != _settings.LabelName).Select((str)=>str.Name).ToList(),
```
Comment (Dmitry-A) [inline src/Microsoft.ML.CodeGen/CodeGenerator/CSharp/CodeGenerator.cs]: Looks to me you're picking features from Columns here, so the lower case features var name is actually deceiving, the Features one will have just features.

File: src/Microsoft.ML.CodeGen/CodeGenerator/CSharp/CodeGeneratorSettings.cs [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+
+        public string OutputName { get; set; }
+
+        public string OutputBaseDir { get; set; }
+
+        public string TrainDataset { get; set; }
+
+        public string TestDataset { get; set; }
+
+        public enum GenerateTarget
```
Comment (Dmitry-A) [inline src/Microsoft.ML.CodeGen/CodeGenerator/CSharp/CodeGeneratorSettings.cs]: Hmm I'm surprised StyleCop would let you have enum intertwined with class members -- is this no longer a style guideline to have them above members?  It makes sense to have it close to the member that uses it but I think most people expect enums to be listed in their own section.

File: src/Microsoft.ML.CodeGen/Microsoft.ML.CodeGenerator.csproj [hunk 11]
```diff
@@ -0,0 +1,114 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFramework>netstandard2.0</TargetFramework>
+    <LangVersion>7.3</LangVersion>
+    <PackageId>Microsoft.ML.CodeGenerator</PackageId>
+    <RootNamespace>Microsoft.ML.CodeGenerator</RootNamespace>
```
Comment (Dmitry-A) [inline src/Microsoft.ML.CodeGen/Microsoft.ML.CodeGenerator.csproj]: This is not needed, I understand it's a copy from somewhere but I had to remove these when merging features/automl code into master

File: src/Microsoft.ML.CodeGen/Microsoft.ML.CodeGenerator.csproj [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+      <Generator>TextTemplatingFilePreprocessor</Generator>
+      <LastGenOutput>PredictProgram.cs</LastGenOutput>
+    </None>
+    <None Update="Templates\Console\PredictProject.tt">
+      <Generator>TextTemplatingFilePreprocessor</Generator>
+      <LastGenOutput>PredictProject.cs</LastGenOutput>
+    </None>
+  </ItemGroup>
+
+  <ItemGroup>
```
Comment (Dmitry-A) [inline src/Microsoft.ML.CodeGen/Microsoft.ML.CodeGenerator.csproj]: Is this needed?  Look to me project would build without it, I suspect it's something that VS puts there by default.

Comment (Dmitry-A) [reply src/Microsoft.ML.CodeGen/Microsoft.ML.CodeGenerator.csproj]: You could also ask Eric Erhardt about this.

File: src/Microsoft.ML.CodeGen/Templates/Console/ConsumeModel.tt [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using Microsoft.ML;
+using <#= Namespace #>.Model;
+using <#= Namespace #>.Model.DataModels;
+
+namespace <#= Namespace #>.Model
+{
+	public class ConsumeModel
```
Comment (Dmitry-A) [inline src/Microsoft.ML.CodeGen/Templates/Console/ConsumeModel.tt]: "Consume" is a strange choice of words for what this does.  Inference or Prediction would be a lot more descriptive.  I couldn't guess what it does until I looked up documentation.  Do you know who came up with this wording?

File: src/Microsoft.ML.CodeGen/Templates/Console/ModelBuilder.cs [hunk 14]
```diff
@@ -390,13 +384,35 @@ public static string GetAbsolutePath(string relativePath)
...
 public bool CacheBeforeTrainer {get;set;}
 public IList<string> PostTrainerTransforms {get;set;}
+public CSharp.CodeGeneratorSettings.GenerateTarget Target {get;set;}
+
+
+void CLI_Annotation()
+{
+this.Write(@"//*****************************************************************************************
+//*                                                                                       *
+//* This is an auto-generated file by Microsoft ML.NET CLI (Command-Line Interface) tool. *
```
Comment (Dmitry-A) [inline src/Microsoft.ML.CodeGen/Templates/Console/ModelBuilder.cs]: Nit: wording.  "This file has been auto-generated by" or something like that.

File: src/Microsoft.ML.CodeGen/Templates/Console/ModelBuilder.cs [hunk 15]
```diff
@@ -390,13 +384,35 @@ public static string GetAbsolutePath(string relativePath)
...
+//*****************************************************************************************
+");
+
+
+}
+
+
+void MB_Annotation()
+{
+this.Write("// This is an auto-generated file by ML.NET Model Builder. \r\n");
```
Comment (Dmitry-A) [inline src/Microsoft.ML.CodeGen/Templates/Console/ModelBuilder.cs]: Same here, wording

File: src/Microsoft.ML.CodeGen/Templates/Console/ModelInputClass.cs [hunk 16]
```diff
@@ -47,13 +44,35 @@ namespace  ");
...
 public IList<string> ClassLabels {get;set;}
 public string Namespace {get;set;}
+public CSharp.CodeGeneratorSettings.GenerateTarget Target {get;set;}
+
+
+void CLI_Annotation()
+{
+this.Write(@"//*****************************************************************************************
+//*                                                                                       *
+//* This is an auto-generated file by Microsoft ML.NET CLI (Command-Line Interface) tool. *
```
Comment (Dmitry-A) [inline src/Microsoft.ML.CodeGen/Templates/Console/ModelInputClass.cs]: Looks to me this needs to be moved out into a function, somehting like AddAnnotation().  That uses correct wording :)

Comment (LittleLittleCloud) [reply src/Microsoft.ML.CodeGen/Templates/Console/ModelInputClass.cs]: this is auto-generated by text template `ModelInputClass.tt`, so no need to worry about

File: src/Microsoft.ML.CodeGen/Templates/Console/ModelInputClass.cs [hunk 17]
```diff
@@ -47,13 +44,35 @@ namespace  ");
...
+//*                                                                                       *
+//*****************************************************************************************
+");
+
+
+}
+
+
+void MB_Annotation()
```
Comment (Dmitry-A) [inline src/Microsoft.ML.CodeGen/Templates/Console/ModelInputClass.cs]: same here

File: src/Microsoft.ML.CodeGen/Templates/Console/ModelOutputClass.cs [hunk 18]
```diff
@@ -60,13 +56,35 @@ namespace  ");
 public string TaskType {get;set;}
 public string PredictionLabelType {get;set;}
 public string Namespace {get;set;}
+public CSharp.CodeGeneratorSettings.GenerateTarget Target {get;set;}
+
+
+void CLI_Annotation()
```
Comment (Dmitry-A) [inline src/Microsoft.ML.CodeGen/Templates/Console/ModelOutputClass.cs]: A third copy of the same exact method?

File: src/Microsoft.ML.CodeGen/Templates/Console/ModelProject.cs [hunk 19]
```diff
@@ -17,17 +17,17 @@ namespace Microsoft.ML.CLI.Templates.Console
...
     {
         /// <summary>
         /// Create the template output
         /// </summary>
         public virtual string TransformText()
         {
-            this.Write("<Project Sdk=\"Microsoft.NET.Sdk\">\r\n\r\n  <PropertyGroup>\r\n    <TargetFramework>netc" +
-                    "oreapp2.1</TargetFramework>\r\n  </PropertyGroup>\r\n  <ItemGroup>\r\n    <PackageRefe" +
-                    "rence Include=\"Microsoft.ML\" Version=\"1.3.1\" />\r\n");
+            this.Write("<Project Sdk=\"Microsoft.NET.Sdk\">\r\n\r\n  <PropertyGroup>\r\n    <TargetFramework>nets" +
```
Comment (Dmitry-A) [inline src/Microsoft.ML.CodeGen/Templates/Console/ModelProject.cs]: TargetFramework probably needs to be a global const somewhere.

File: src/Microsoft.ML.CodeGen/Templates/Console/ModelBuilder.cs (hunk 14 above)
Comment (LittleLittleCloud) [reply src/Microsoft.ML.CodeGen/Templates/Console/ModelBuilder.cs]: @briacht  for you

File: src/Microsoft.ML.CodeGen/Templates/Console/ModelInputClass.cs (hunk 17 above)
Comment (LittleLittleCloud) [reply src/Microsoft.ML.CodeGen/Templates/Console/ModelInputClass.cs]: resolve due to replication

File: src/Microsoft.ML.CodeGen/Templates/Console/ModelBuilder.cs (hunk 15 above)
Comment (LittleLittleCloud) [reply src/Microsoft.ML.CodeGen/Templates/Console/ModelBuilder.cs]: resolve due to replication

File: src/Microsoft.ML.CodeGen/Templates/Console/ConsumeModel.tt (hunk 13 above)
Comment (LittleLittleCloud) [reply src/Microsoft.ML.CodeGen/Templates/Console/ConsumeModel.tt]: @briacht

File: src/Microsoft.ML.CodeGen/Microsoft.ML.CodeGenerator.csproj (hunk 11 above)
Comment (LittleLittleCloud) [reply src/Microsoft.ML.CodeGen/Microsoft.ML.CodeGenerator.csproj]: removed

File: src/Microsoft.ML.CodeGen/CodeGenerator/CSharp/CodeGeneratorSettings.cs (hunk 10 above)
Comment (LittleLittleCloud) [reply src/Microsoft.ML.CodeGen/CodeGenerator/CSharp/CodeGeneratorSettings.cs]: move `GenerateTarget` out of `CodeGeneratorSettings` Class

File: src/Microsoft.ML.CodeGen/Microsoft.ML.CodeGenerator.csproj (hunk 12 above)
Comment (LittleLittleCloud) [reply src/Microsoft.ML.CodeGen/Microsoft.ML.CodeGenerator.csproj]: It's auto-added when you add a text template file into project... So I guess it's just a method to inform Visual Studio that this project has tt file, so Visual Studio can treat the generated CSharp file differently ( see the UI in file explorer for example). 

In that case I guess we can remove that item in mlnet.csproj, cause it doesn't include any tt template any more

Comment (LittleLittleCloud) [reply src/Microsoft.ML.CodeGen/Microsoft.ML.CodeGenerator.csproj]: @eerhardt Please correct me if I'm wrong

Comment (eerhardt) [reply src/Microsoft.ML.CodeGen/Microsoft.ML.CodeGenerator.csproj]: Unfortunately, yes when you use `.tt` files, this Service will keep being added to your project, even if you try to remove it.

---
In reply to: [315343898](https://github.com/dotnet/machinelearning/pull/4043#discussion_r315343898) [](ancestors = 315343898)

**Comment by eerhardt** [comment]
> <?xml version="1.0" encoding="utf-8" ?>

I don't think you need an `NLog.config` file for a library. That should only be at the app level, I think.

---
Refers to: src/Microsoft.ML.CodeGen/NLog.config:1 in a847f05. [](commit_id = a847f054722e951e81fa2473a18da8455f5c7a07, deletion_comment = False)

File: src/Microsoft.ML.CodeGen/Templates/Console/PredictProgram.tt [hunk 20]
```diff
@@ -0,0 +1,92 @@
+﻿<#@ template language="C#" linePragmas="false" #>
```
Comment (Dmitry-A) [inline src/Microsoft.ML.CodeGen/Templates/Console/PredictProgram.tt]: https://github.com/dotnet/machinelearning/issues/4131 for removing generated .cs files

**Comment by srsaggam** [comment]
Could run a smoke test with different datasets per task type. binary, multi and regression. and check the generated projects can compile and run and give meaningful results before closing in on this PR

File: src/Microsoft.ML.CodeGen/CodeGenerator/CSharp/CodeGenerator.cs (hunk 9 above)
Comment (LittleLittleCloud) [reply src/Microsoft.ML.CodeGen/CodeGenerator/CSharp/CodeGenerator.cs]: resolved

File: src/Microsoft.ML.CodeGen/CodeGenerator/CSharp/CodeGenerator.cs (hunk 8 above)
Comment (LittleLittleCloud) [reply src/Microsoft.ML.CodeGen/CodeGenerator/CSharp/CodeGenerator.cs]: resolved
features => featuresList

**Comment by LittleLittleCloud** [comment]
> Could you run a smoke test with different datasets per task type. binary, multi and regression. and check the generated projects can compile and run and give meaningful results before closing in on this PR

- [x] binary
- [x] multi
- [x] regression

File: src/Microsoft.ML.CodeGen/Templates/Console/ModelBuilder.cs (hunk 14 above)
Comment (briacht) [reply src/Microsoft.ML.CodeGen/Templates/Console/ModelBuilder.cs]: Change to: 
MB: "This file was auto-generated by ML.NET Model Builder."

CLI: "This file was auto-generated by ML.NET CLI (Command-Line Interface) tool."

Comment (LittleLittleCloud) [reply src/Microsoft.ML.CodeGen/Templates/Console/ModelBuilder.cs]: K

File: src/Microsoft.ML.CodeGen/Templates/Console/ConsumeModel.tt (hunk 13 above)
Comment (Dmitry-A) [reply src/Microsoft.ML.CodeGen/Templates/Console/ConsumeModel.tt]: One more related question: there's already a prediction project, why do we need a second one?

Comment (briacht) [reply src/Microsoft.ML.CodeGen/Templates/Console/ConsumeModel.tt]: Through user testing and reported issues we found that users struggled with the last (Code) step in Model Builder e.g. they could not figure out how to easily consume the model with the code that they were given. Additionally, the consume code became very repetitive, so we are providing users with a separate Predict method so that they only have to add namespaces, create input, and then predict.

We have used the term “consume model” in our getting started experiences / when talking about loading and using the ML model for predictions, so for now we will leave it as is for this release and then do further user testing to validate / invalidate.

File: src/Microsoft.ML.CodeGen/CodeGenerator/CSharp/CodeGeneratorSettings.cs [hunk 21]
```diff
@@ -0,0 +1,39 @@
...
+// See the LICENSE file in the project root for more information.
+using Microsoft.ML.AutoML;
+
+namespace Microsoft.ML.CodeGenerator.CSharp
+{
+    public class CodeGeneratorSettings
+    {
+        public CodeGeneratorSettings()
+        {
+            // set default value
```
Comment (justinormont) [inline src/Microsoft.ML.CodeGen/CodeGenerator/CSharp/CodeGeneratorSettings.cs]: Comment style in ML.NET is to begin comment with an uppercase character
```suggestion
            // Set default value
```

File: src/Microsoft.ML.CodeGen/Templates/Console/ModelOutputClass.cs [hunk 22]
```diff
@@ -17,24 +17,20 @@ namespace Microsoft.ML.CLI.Templates.Console
...
-//*****************************************************************************************
-
-using System;
-using Microsoft.ML.Data;
-
-namespace  ");
+if(Target == CSharp.GenerateTarget.Cli){ 
+CLI_Annotation();
+ } else if(Target == CSharp.GenerateTarget.ModelBuilder){ 
+MB_Annotation();
```
Comment (justinormont) [inline src/Microsoft.ML.CodeGen/Templates/Console/ModelOutputClass.cs]: Instead of having a `if CLI ... else if ModelBuilder, ... else if NextCreatedCallerOfCodeGen`, should we have a parameter which takes set the Annotation?

Our default can be something like: "This is an auto-generated file by ML.NET AutoML". 

This will let future callers set their own annotation, and makes it so we're not listing the set of callers to an API.

---

In addition, the C# style in the ML.NET repo would for the current method be:
```C#
if (Target == CSharp.GenerateTarget.Cli)
{ 
  CLI_Annotation();
} 
else if (Target == CSharp.GenerateTarget.ModelBuilder)
{ 
  MB_Annotation();
}
```

I'm uncertain why StyleCode didn't flag this.

File: src/Microsoft.ML.CodeGen/CodeGenerator/CSharp/CodeGeneratorSettings.cs (hunk 21 above)
Comment (LittleLittleCloud) [reply src/Microsoft.ML.CodeGen/CodeGenerator/CSharp/CodeGeneratorSettings.cs]: Thanks for pointing out

File: src/Microsoft.ML.CodeGen/CodeGenerator/CSharp/CodeGeneratorSettings.cs [hunk 23]
```diff
@@ -0,0 +1,39 @@
...
+
+        public string TestDataset { get; set; }
+
+        public GenerateTarget Target { get; set; }
+
+        internal TaskKind MlTask { get; set; }
+
+    }
+
+    public enum GenerateTarget
```
Comment (justinormont) [inline src/Microsoft.ML.CodeGen/CodeGenerator/CSharp/CodeGeneratorSettings.cs]: This seem to be used only to print the right annotation for the platform. I might recommend just having a string field here for the annotation text.

This would address my other comment: https://github.com/dotnet/machinelearning/pull/4043/files#r317240647

File: src/Microsoft.ML.CodeGen/Templates/Console/Annotation.txt [hunk 24]
```diff
@@ -0,0 +1 @@
+﻿
```
Comment (justinormont) [inline src/Microsoft.ML.CodeGen/Templates/Console/Annotation.txt]: What's the purpose of this empty file?

File: src/Microsoft.ML.CodeGen/Templates/Console/ModelInputClass.cs (hunk 17 above)
Comment (justinormont) [reply src/Microsoft.ML.CodeGen/Templates/Console/ModelInputClass.cs]: #resolved

File: src/Microsoft.ML.CodeGen/CodeGenerator/CSharp/CodeGeneratorSettings.cs (hunk 21 above)
Comment (LittleLittleCloud) [reply src/Microsoft.ML.CodeGen/CodeGenerator/CSharp/CodeGeneratorSettings.cs]: #resolved

File: src/Microsoft.ML.CodeGen/Templates/Console/Annotation.txt (hunk 24 above)
Comment (LittleLittleCloud) [reply src/Microsoft.ML.CodeGen/Templates/Console/Annotation.txt]: Oh it's generated by T4... I will remove that

File: src/Microsoft.ML.CodeGen/Templates/Console/ModelProject.tt [hunk 25]
```diff
@@ -6,7 +6,7 @@
 <Project Sdk="Microsoft.NET.Sdk">
 
   <PropertyGroup>
-    <TargetFramework>netcoreapp2.1</TargetFramework>
+    <TargetFramework>netstandard2.0</TargetFramework>
```
Comment (justinormont) [inline src/Microsoft.ML.CodeGen/Templates/Console/ModelProject.tt]: Any reason this moved from core to standard? Does this allow the generated project to exist in a wider set of projects?

Comment (LittleLittleCloud) [reply src/Microsoft.ML.CodeGen/Templates/Console/ModelProject.tt]: Because ML.Net target netstandard 2.0 and we want to keep in pace with that. 
@JakeRadMSFT is that correct?

Comment (JakeRadMSFT) [reply src/Microsoft.ML.CodeGen/Templates/Console/ModelProject.tt]: Yeah this makes it so .NET Framework and .NET Core projects can reference the model project.
